File: feature/divkit-custom-feature/src/androidMain/kotlin/ru/pyroman/news/feature/divkit/custom/DivkitCustomModule.kt
package ru.pyroman.news.feature.divkit.custom

import org.kodein.di.DI
import org.kodein.di.instance
import ru.pyroman.news.common.core.di.provider
import ru.pyroman.news.common.core.platform.di.contextInstance
import ru.pyroman.news.feature.divkit.custom.search.SearchDivCustomViewAdapter
import ru.pyroman.news.feature.search.searchModule

val divkitCustomModule = DivkitCustomModuleAssembler.apply {
    assembleModule(
        childModules = listOf(
            searchModule,
        ),
        bridge = DivkitCustomModuleBridgeImpl(),
    )
}.instance

private class DivkitCustomModuleBridgeImpl : DivkitCustomModuleBridge() {

    override fun provideNativeInstances(builder: DI.Builder) {
        with(builder) {
            provider {
                SearchDivCustomViewAdapter(
                    context = contextInstance(),
                    searchPresenterFactory = instance(),
                )
            }

            provider<DivCustomViewAdaptersProvider> {
                DivCustomViewAdaptersProviderImpl(
                    searchDivCustomViewAdapter = instance(),
                )
            }
        }
    }
}

File: iosApp/News/Modules/DivkitCustomFeature/Sources/DivkitCustomModule.swift
import CommonUmbrella
import SearchFeature

public let divkitCustomModule = makeDivkitCustomModule()

private func makeDivkitCustomModule() -> DIModule {
    DivkitCustomModuleAssembler.shared.assembleModule(
        childModules: [
            searchModule
        ],
        bridge: DivkitCustomModuleBridgeImpl()
    )
    return DivkitCustomModuleAssembler.shared.instance
}

private class DivkitCustomModuleBridgeImpl: DivkitCustomModuleBridge {
    
    override func provideNativeInstances(builder: any DIBuilder) {
        
        DIExtensionsKt.taggedProvider(builder, tag: "SearchDivCustomBlockFactory") { directDIAware in
            SearchDivCustomBlockFactory(
                observeSearchInputUseCase: DIExtensionsKt.taggedInstance(directDIAware, tag: "ObserveSearchInputUseCase") as! ObserveSearchInputUseCase
            )
        }
        
        DIExtensionsKt.taggedProvider(builder, tag: "DivCustomBlockFactory") { directDIAware in
            DivCustomBlockFactoryImpl(
                searchDivCustomBlockFactory: DIExtensionsKt.taggedInstance(directDIAware, tag: "SearchDivCustomBlockFactory") as! SearchDivCustomBlockFactory
            )
        }
    }
}
